### ISSUE TITLE
jsx-curly-brace-presence ignores strings with "{" or "}" in the text

### ISSUE THREAD

**Issue by chestozo** [yannickcr/eslint-plugin-react#2694]
I have this setup in `.eslintrc`
```
"react/jsx-curly-brace-presence": [
  "error",
  {
    "props": "never",
    "children": "ignore"
  }
]
```

in this case I get
```jsx
<Box mb={'1rem'} />    // ERROR - all good here I expect this error
<Box mb={'1rem {}'} /> // NO ERROR - this looks like a bug (should be ERROR too)
```

**Comment by jzabala** [comment]
Hi @chestozo, from what I see your case is covered at the end of the edge cases in the rule [docs][docs].

The rationale is that in some cases it can make the code less readable:

**With:**
```jsx
<style type="text/css">{'.main { margin-top: 0; }'}</style>;
```
**Without:**
```jsx
<style type="text/css">'.main { margin-top: 0; }'</style>;
```

[docs]: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-curly-brace-presence.md#edge-cases

**Comment by ljharb** [comment]
In that latter case, that seems like it'd actually be invalid.

@chestozo the only way to have curly braces show up in your string is if they're encoded or in quotes; i think the rule is working as intended here.

**Comment by chestozo** [comment]
> @chestozo the only way to have curly braces show up in your string is if they're encoded or in quotes; i think the rule is working as intended here.

would you agree that `' hello {name}'` is a totally valid string that does not actually require
- encoding
- quoting
?

if yes this means it is still a bug ¯\_(ツ)_/¯

**Comment by chestozo** [comment]
@jzabala not really
my case is for attribute contents not for children

**Comment by ljharb** [comment]
@chestozo I think that it seems like a programmer bug, which is why the linter rule should be forcing it to be a string inside jsx interpolation (`{ }`), or, to use HTML encoding - that way it's very explicit that it's intentional.

**Comment by chestozo** [comment]
@ljharb 
it is not a bug at all
here is a life example of such a string when react-intl `FormattedMessage` is used:
```tsx
<FormattedMessage
  id='confirmAction.enterCodeMessage'
  defaultMessage='Enter code from {type}'
  values={{ type: 'email' }}
/>
```

**Comment by ljharb** [comment]
ahhh! I agree that in a *prop value*, you should not be getting a warning from this rule. Sorry I misunderstood your OP.

**Comment by jzabala** [comment]
> ahhh! I agree that in a _prop value_, you should not be getting a warning from this rule. Sorry I misunderstood your OP.

Actually this rule has two tests that specifically guard agains this (disallowed JSX text characters):

https://github.com/yannickcr/eslint-plugin-react/blob/e3e767bd041988d9acb7713874c0632c68408347/tests/lib/rules/jsx-curly-brace-presence.js#L252

https://github.com/yannickcr/eslint-plugin-react/blob/e3e767bd041988d9acb7713874c0632c68408347/tests/lib/rules/jsx-curly-brace-presence.js#L248

Would them happen to be allowed values only on props?

**Comment by ljharb** [comment]
If the OP’s test cases pass, we can close the issue by adding them. Indeed if they’re prop values they should not be warned on.

**Comment by jzabala** [comment]
> If the OP’s test cases pass, we can close the issue by adding them. Indeed if they’re prop values they should not be warned on.

I opened a PR with the changes and moved the tests.
